xds/src/test/java/io/grpc/xds/XdsServerTestHelper.java

```diff
@@ -230,27 +248,61 @@ public <T extends ResourceUpdate> void cancelXdsResourceWatch(XdsResourceType<T>
     }
 
     @Override
-    public void shutdown() {
+    public synchronized void shutdown() {
       shutdown = true;
     }
 
     @Override
-    public boolean isShutDown() {
+    public synchronized boolean isShutDown() {
       return shutdown;
     }
 
+    public void awaitRds() throws InterruptedException, TimeoutException {
+      if (!rdsCount.await(WAIT_TIMEOUT.toMillis(), TimeUnit.MILLISECONDS)) {
+        throw new TimeoutException("Timeout " + WAIT_TIMEOUT + " waiting for RDSs");
+      }
+    }
+
+    public void setExpectedRdsCount(int count) {
+      rdsCount = new CountDownLatch(count);
+    }
+
+    private void execute(Runnable action) {
+      // This method ensures that all watcher updates:
+      // - Happen after the server started watching LDS.
+      // - Are executed within the sync context of the server.
+      //
+      // Note that this doesn't guarantee that any of the RDS watchers are created.
+      // Tests should use setExpectedRdsCount(int) and awaitRds() for that.
+      if (ldsResource == null) {
+        throw new IllegalStateException("xDS resource update after watcher cancel");
+      }
+      try {
+        ldsResource.get(WAIT_TIMEOUT.toMillis(), TimeUnit.MILLISECONDS);
+      } catch (ExecutionException | TimeoutException e) {
+        throw new RuntimeException("Can't resolve LDS resource name in " + WAIT_TIMEOUT, e);
+      } catch (InterruptedException e) {
+        Thread.currentThread().interrupt();
+        throw new RuntimeException(e);
+      }
+      assertThat(serverExecutor).isNotNull();
```

Comment:
This looks like it is being helpful, but it isn't. If `serverExecutor` is null, then the next line will already throw. In addition, `serverExecutor` is volatile, so there should be no assumption it has the same value on this line and the next.